File: demo/src/main/java/com/example/demo/repository/ProductsRepository.java
package com.example.demo.repository;

import com.example.demo.entity.Products;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.List;
import java.util.Map;

public interface ProductsRepository extends JpaRepository<Products,Integer>
{

    @Query(value =
            "SELECT p.id, p.name, p.slug, p.short_description, p.full_description, " +
                    "p.created_at, p.updated_at, p.seller_id, p.brand_id, p.status, " +
                    "s.id as seller_id, s.name as seller_name, s.description as seller_description, " +
                    "s.logo_url as seller_logo_url, s.rating as seller_ratings, " +
                    "s.contact_email as seller_contact_email, s.created_at as seller_created_at, " +
                    "b.id as brand_id, b.name as brand_name, b.slug as brand_slug, " +
                    "b.description as brand_description, b.logo_url as brand_logo_url, " +
                    "b.website as brand_website, " +
                    "pd.id as product_detail_id, pd.product_id as p_detail_product_id, pd.weight, pd.dimensions, " +
                    "pd.materials, pd.country_of_origin, " +
                    "pd.warranty_info, pd.care_instructions, " +
                    "pd.additional_info, " +
                    "pp.id as product_price_id, pp.product_id as product_price_product_id, pp.base_price, pp.cost_price, pp.currency, pp.tax_rate, " +
                    "pog.id as option_group_id, pog.product_id as option_group_product_id, " +
                    "pog.name as option_group_name, pog.display_order as option_group_display_order, " +
                    "po.id as option_id, po.option_group_id as option_option_group_id, " +
                    "po.name as option_name, po.additional_price, " +
                    "po.sku, po.stock, po.display_order as option_display_order, " +
                    "pi.id as image_id, pi.product_id as image_product_id, " +
                    "pi.url as image_url, pi.alt_text, " +
                    "pi.is_primary, pi.display_order as image_display_order, " +
                    "pi.option_id, " +
                    "pc.id as p_category_id, pc.product_id as p_category_product_id, " +
                    "pc.category_id as p_category_category_id, pc.is_primary as category_is_primary, " +
                    "c.id as category_id, c.name as category_name, " +
                    "c.slug as category_slug, c.description as category_description, " +
                    "c.parent_id, c.level, c.image_url as category_image_url, " +
                    "pt.id as p_tag_id, pt.product_id as p_tag_product_id, " +
                    "pt.tag_id as p_tag_tag_id, " +
                    "t.id as tag_id, t.name as tag_name, " +
                    "t.slug as tag_slug " +
                    "FROM products p " +
                    "LEFT OUTER JOIN sellers s ON p.seller_id = s.id " +
                    "LEFT OUTER JOIN brands b ON p.brand_id = b.id " +
                    "LEFT OUTER JOIN product_details pd ON p.id = pd.product_id " +
                    "LEFT OUTER JOIN product_prices pp ON p.id = pp.product_id " +
                    "LEFT OUTER JOIN product_option_groups pog ON p.id = pog.product_id " +
                    "LEFT OUTER JOIN product_options po ON pog.id = po.option_group_id " +
                    "LEFT OUTER JOIN product_images pi ON p.id = pi.product_id " +
                    "LEFT OUTER JOIN product_categories pc ON p.id = pc.product_id " +
                    "LEFT OUTER JOIN categories c ON pc.category_id = c.id " +
                    "LEFT OUTER JOIN product_tags pt ON p.id = pt.product_id " +
                    "LEFT OUTER JOIN tags t ON pt.tag_id = t.id",
            nativeQuery = true)
    List<Map<String, Object>> findAllProductDetails();

    @Query(value =
            "SELECT p.id, p.name, p.slug, p.short_description, p.full_description, " +
                    "p.created_at, p.updated_at, p.seller_id, p.brand_id, p.status, " +
                    "s.id as seller_id, s.name as seller_name, s.description as seller_description, " +
                    "s.logo_url as seller_logo_url, s.rating as seller_ratings, " +
                    "s.contact_email as seller_contact_email, s.created_at as seller_created_at, " +
                    "b.id as brand_id, b.name as brand_name, b.slug as brand_slug, " +
                    "b.description as brand_description, b.logo_url as brand_logo_url, " +
                    "b.website as brand_website, " +
                    "pd.id as product_detail_id, pd.product_id as p_detail_product_id, pd.weight, pd.dimensions, " +
                    "pd.materials, pd.country_of_origin, " +
                    "pd.warranty_info, pd.care_instructions, " +
                    "pd.additional_info, " +
                    "pp.id as product_price_id, pp.product_id as product_price_product_id, pp.base_price, pp.cost_price, pp.currency, pp.tax_rate, " +
                    "pog.id as option_group_id, pog.product_id as option_group_product_id, " +
                    "pog.name as option_group_name, pog.display_order as option_group_display_order, " +
                    "po.id as option_id, po.option_group_id as option_option_group_id, " +
                    "po.name as option_name, po.additional_price, " +
                    "po.sku, po.stock, po.display_order as option_display_order, " +
                    "pi.id as image_id, pi.product_id as image_product_id, " +
                    "pi.url as image_url, pi.alt_text, " +
                    "pi.is_primary, pi.display_order as image_display_order, " +
                    "pi.option_id, " +
                    "pc.id as p_category_id, pc.product_id as p_category_product_id, " +
                    "pc.category_id as p_category_category_id, pc.is_primary as category_is_primary, " +
                    "c.id as category_id, c.name as category_name, " +
                    "c.slug as category_slug, c.description as category_description, " +
                    "c.parent_id, c.level, c.image_url as category_image_url, " +
                    "pt.id as p_tag_id, pt.product_id as p_tag_product_id, " +
                    "pt.tag_id as p_tag_tag_id, " +
                    "t.id as tag_id, t.name as tag_name, " +
                    "t.slug as tag_slug " +
                    "FROM products p " +
                    "LEFT OUTER JOIN sellers s ON p.seller_id = s.id " +
                    "LEFT OUTER JOIN brands b ON p.brand_id = b.id " +
                    "LEFT OUTER JOIN product_details pd ON p.id = pd.product_id " +
                    "LEFT OUTER JOIN product_prices pp ON p.id = pp.product_id " +
                    "LEFT OUTER JOIN product_option_groups pog ON p.id = pog.product_id " +
                    "LEFT OUTER JOIN product_options po ON pog.id = po.option_group_id " +
                    "LEFT OUTER JOIN product_images pi ON p.id = pi.product_id " +
                    "LEFT OUTER JOIN product_categories pc ON p.id = pc.product_id " +
                    "LEFT OUTER JOIN categories c ON pc.category_id = c.id " +
                    "LEFT OUTER JOIN product_tags pt ON p.id = pt.product_id " +
                    "LEFT OUTER JOIN tags t ON pt.tag_id = t.id " +
                    "WHERE p.id = :productId",
            nativeQuery = true)
    List<Map<String, Object>> findProductDetailsById(@Param("productId") Integer productId);
}

File: src/modules/products/products.repository.ts
import { Injectable } from '@nestjs/common';
import { DrizzleService } from '~/database/drizzle.service';
import { and, eq, like, lte, or, sql, gte } from 'drizzle-orm';
import { getOrderBy } from '~/common/utils/drizzle-helpers';

import {
  products as productsSchema,
  productCategories as productCategoriesSchema,
  productImages as productImagesSchema,
  productOptionGroups as productOptionGroupsSchema,
  productOptions as productOptionsSchema,
  productPrices as productPricesSchema,
  productDetails as productDetailsSchema,
  productTags as productTagsSchema,
} from '~/database/schema';
import {
  CreateProductRequestDto,
  GetProductsRequestDto,
  UpdateProductRequestDto,
  CraeteProductResponseData,
  UpdateProductResponseData,
} from './dto/product.dto';
import { ProductWithRelations } from './entities/product.entity';
import {
  CreateProductOptionRequestDto,
  CreateProductOptionResponseData,
  CreateProductOptionResponseDto,
  UpdateProductOptionRequestDto,
  UpdateProductOptionResponseData,
  UpdateProductOptionResponseDto,
} from './dto/product-option.dto';
import {
  CreateProductImageRequestDto,
  CreateProductImageResponseData,
  CreateProductImageResponseDto,
} from './dto/product-image.dto';

@Injectable()
export class ProductsRepository {
  constructor(private readonly drizzleService: DrizzleService) {}

  async getProducts(
    params: GetProductsRequestDto,
  ): Promise<ProductWithRelations[]> {
    const {
      page = 1,
      per_page: perPage = 10,
      sort = 'created_at:desc',
      ...filters
    } = params;
    const offset = (page - 1) * perPage;

    const where = this.buildProductQueryConditions(filters);

    return this.drizzleService.db.query.products.findMany({
      with: this.getProductListRelations(),
      where,
      orderBy: getOrderBy('product', sort),
      limit: perPage,
      offset,
    });
  }

  async getProductsCount(params: GetProductsRequestDto): Promise<number> {
    const where = this.buildProductQueryConditions(params);

    const result = await this.drizzleService.db
      .select({ count: sql`count(*)` })
      .from(productsSchema)
      .where(where)
      .then((result) => Number(result[0].count));

    return result;
  }

  async getProduct(id: number): Promise<ProductWithRelations> {
    const product = await this.drizzleService.db.query.products.findFirst({
      with: this.getProductDetailRelations(),
      where: eq(productsSchema.id, id),
    });

    return product;
  }

  async createProduct(
    data: CreateProductRequestDto,
  ): Promise<CraeteProductResponseData> {
    const {
      categories,
      optionGroups,
      images,
      detail,
      price,
      tags,
      ...productData
    } = data;

    const [product] = await this.drizzleService.db.transaction(async (tx) => {
      const [newProduct] = await tx
        .insert(productsSchema)
        .values({
          id: undefined,
          name: productData.name || '',
          slug: productData.slug || '',
          shortDescription: productData.shortDescription || '',
          fullDescription: productData.fullDescription || '',
          status: productData.status || 'ACTIVE',
          sellerId: productData.sellerId,
          brandId: productData.brandId,
          createdAt: new Date(),
          updatedAt: new Date(),
        })
        .returning();

      if (detail) {
        await tx.insert(productDetailsSchema).values({
          id: undefined,
          productId: newProduct.id,
          weight: String(detail.weight),
          dimensions: detail.dimensions,
          materials: detail.materials,
          countryOfOrigin: detail.countryOfOrigin,
          warrantyInfo: detail.warrantyInfo,
          careInstructions: detail.careInstructions,
          additionalInfo: detail.additionalInfo,
        });
      }

      if (price) {
        await tx.insert(productPricesSchema).values({
          id: undefined,
          productId: newProduct.id,
          basePrice: String(price.basePrice),
          salePrice: price.salePrice ? String(price.salePrice) : undefined,
          costPrice: price.costPrice ? String(price.costPrice) : undefined,
          currency: price.currency,
          taxRate: price.taxRate ? String(price.taxRate) : undefined,
        });
      }

      if (categories?.length) {
        await tx.insert(productCategoriesSchema).values(
          categories.map((cat: any) => ({
            id: undefined,
            productId: newProduct.id,
            categoryId: cat.categoryId,
            isPrimary: cat.isPrimary,
          })),
        );
      }

      if (optionGroups?.length) {
        for (const group of optionGroups) {
          const [newGroup] = await tx
            .insert(productOptionGroupsSchema)
            .values({
              id: undefined,
              name: group.name,
              productId: newProduct.id,
              displayOrder: group.displayOrder,
            })
            .returning();

          if (group.options?.length) {
            await tx.insert(productOptionsSchema).values(
              group.options.map((option: any) => ({
                id: undefined,
                optionGroupId: newGroup.id,
                name: option.name,
                additionalPrice: option.additionalPrice,
                sku: option.sku,
                stock: option.stock,
                displayOrder: option.displayOrder,
              })),
            );
          }
        }
      }

      if (images?.length) {
        await tx.insert(productImagesSchema).values(
          images.map((img: any) => ({
            id: undefined,
            productId: newProduct.id,
            url: img.url,
            altText: img.altText,
            isPrimary: img.isPrimary,
            displayOrder: img.displayOrder,
            optionId: img.optionId,
          })),
        );
      }

      if (tags?.length) {
        await tx.insert(productTagsSchema).values(
          tags.map((tagId: number) => ({
            id: undefined,
            productId: newProduct.id,
            tagId,
          })),
        );
      }

      return [newProduct];
    });

    return product;
  }

  async updateProduct(
    id: number,
    data: UpdateProductRequestDto,
  ): Promise<UpdateProductResponseData> {
    const {
      categories,
      optionGroups,
      images,
      detail,
      price,
      tags,
      ...productData
    } = data;

    const [product] = await this.drizzleService.db.transaction(async (tx) => {
      const [updatedProduct] = await tx
        .update(productsSchema)
        .set({ ...productData, updatedAt: new Date() })
        .where(eq(productsSchema.id, id))
        .returning();

      if (detail) {
        await tx.insert(productDetailsSchema).values({
          id: undefined,
          productId: updatedProduct.id,
          weight: String(detail.weight),
          dimensions: detail.dimensions,
          materials: detail.materials,
          countryOfOrigin: detail.countryOfOrigin,
          warrantyInfo: detail.warrantyInfo,
          careInstructions: detail.careInstructions,
          additionalInfo: detail.additionalInfo,
        });
      }

      if (price) {
        await tx.insert(productPricesSchema).values({
          id: undefined,
          productId: updatedProduct.id,
          basePrice: String(price.basePrice),
          salePrice: price.salePrice ? String(price.salePrice) : undefined,
          costPrice: price.costPrice ? String(price.costPrice) : undefined,
          currency: price.currency,
          taxRate: price.taxRate ? String(price.taxRate) : undefined,
        });
      }

      if (categories) {
        await tx
          .delete(productCategoriesSchema)
          .where(eq(productCategoriesSchema.productId, id));
        if (categories.length) {
          await tx.insert(productCategoriesSchema).values(
            categories.map((cat: any) => ({
              id: undefined,
              productId: id,
              categoryId: cat.categoryId,
              isPrimary: cat.isPrimary,
            })),
          );
        }
      }

      if (optionGroups) {
        const existingGroups = await tx
          .select()
          .from(productOptionGroupsSchema)
          .where(eq(productOptionGroupsSchema.productId, id));

        for (const group of existingGroups) {
          await tx
            .delete(productOptionsSchema)
            .where(eq(productOptionsSchema.optionGroupId, group.id));
        }

        await tx
          .delete(productOptionGroupsSchema)
          .where(eq(productOptionGroupsSchema.productId, id));

        if (optionGroups.length) {
          for (const group of optionGroups) {
            const [newGroup] = await tx
              .insert(productOptionGroupsSchema)
              .values({
                id: undefined,
                name: group.name,
                productId: id,
                displayOrder: group.displayOrder,
              })
              .returning();

            if (group.options?.length) {
              await tx.insert(productOptionsSchema).values(
                group.options.map((option: any) => ({
                  id: undefined,
                  optionGroupId: newGroup.id,
                  name: option.name,
                  additionalPrice: option.additionalPrice,
                  sku: option.sku,
                  stock: option.stock,
                  displayOrder: option.displayOrder,
                })),
              );
            }
          }
        }
      }

      if (images) {
        await tx
          .delete(productImagesSchema)
          .where(eq(productImagesSchema.productId, id));
        if (images.length) {
          await tx.insert(productImagesSchema).values(
            images.map((img: any) => ({
              id: undefined,
              productId: id,
              url: img.url,
              altText: img.altText,
              isPrimary: img.isPrimary,
              displayOrder: img.displayOrder,
              optionId: img.optionId,
            })),
          );
        }
      }

      if (tags?.length) {
        await tx.insert(productTagsSchema).values(
          tags.map((tagId: number) => ({
            id: undefined,
            productId: updatedProduct.id,
            tagId,
          })),
        );
      }

      return [updatedProduct];
    });

    return product;
  }

  async deleteProduct(id: number): Promise<void> {
    await this.drizzleService.db
      .delete(productsSchema)
      .where(eq(productsSchema.id, id));
  }

  private buildProductQueryConditions = (filters: GetProductsRequestDto) => {
    const conditions = [];

    if (filters.status) {
      conditions.push(eq(productsSchema.status, filters.status));
    }
    if (filters.minPrice) {
      conditions.push(
        gte(productPricesSchema.basePrice, String(filters.minPrice)),
      );
    }
    if (filters.maxPrice) {
      conditions.push(
        lte(productPricesSchema.basePrice, String(filters.maxPrice)),
      );
    }
    if (filters.category?.length) {
      conditions.push(
        sql`${productsSchema.id} IN (
          SELECT product_id 
          FROM ${productCategoriesSchema} 
          WHERE category_id IN (${filters.category})
        )`,
      );
    }
    if (filters.seller) {
      conditions.push(eq(productsSchema.sellerId, filters.seller));
    }
    if (filters.brand) {
      conditions.push(eq(productsSchema.brandId, filters.brand));
    }
    if (filters.search) {
      conditions.push(
        or(
          like(productsSchema.name, `%${filters.search}%`),
          like(productsSchema.shortDescription, `%${filters.search}%`),
        ),
      );
    }

    return conditions.length ? and(...conditions) : undefined;
  };

  private getProductListRelations() {
    return {
      brand: {
        columns: {
          id: true,
          name: true,
          logoUrl: true,
        },
      },
      seller: {
        columns: {
          id: true,
          name: true,
          logoUrl: true,
        },
      },
      images: {
        where: eq(productImagesSchema.isPrimary, true),
        columns: {
          url: true,
          altText: true,
        },
      },
      reviews: {
        columns: {
          rating: true,
        },
      },
      optionGroups: {
        with: {
          options: {
            columns: {
              stock: true,
            },
          },
        },
      },
    };
  }

  private getProductDetailRelations() {
    return {
      seller: {
        columns: {
          id: true,
          name: true,
          logoUrl: true,
          rating: true,
          contactEmail: true,
          contactPhone: true,
        },
      },
      brand: {
        columns: {
          id: true,
          name: true,
          logoUrl: true,
          website: true,
        },
      },
      detail: {
        columns: {
          id: true,
          weight: true,
          dimensions: true,
          materials: true,
          countryOfOrigin: true,
          warrantyInfo: true,
          careInstructions: true,
          additionalInfo: true,
        },
      },
      price: {
        columns: {
          id: true,
          basePrice: true,
          salePrice: true,
          currency: true,
          taxRate: true,
        },
      },
      categories: {
        columns: {
          isPrimary: true,
        },
        with: {
          category: {
            columns: {
              id: true,
              name: true,
              slug: true,
            },
            with: {
              parent: {
                columns: {
                  id: true,
                  name: true,
                  slug: true,
                },
              },
            },
          },
        },
      },
      optionGroups: {
        with: {
          options: {
            columns: {
              id: true,
              name: true,
              stock: true,
              additionalPrice: true,
              sku: true,
              displayOrder: true,
            },
          },
        },
      },
      images: {
        columns: {
          id: true,
          url: true,
          altText: true,
          isPrimary: true,
          displayOrder: true,
          optionId: true,
        },
      },
      tags: {
        with: {
          tag: {
            columns: {
              id: true,
              name: true,
              slug: true,
            },
          },
        },
      },
      reviews: {
        with: {
          user: {
            columns: {
              id: true,
              name: true,
              avatarUrl: true,
            },
          },
        },
      },
    };
  }

  async createProductOption(
    dto: CreateProductOptionRequestDto,
  ): Promise<CreateProductOptionResponseData> {
    const [option] = await this.drizzleService.db
      .insert(productOptionsSchema)
      .values({
        id: undefined,
        optionGroupId: dto.optionGroupId,
        name: dto.name,
        additionalPrice: String(dto.additionalPrice),
        sku: dto.sku,
        stock: dto.stock,
        displayOrder: dto.displayOrder,
      })
      .returning({
        id: productOptionsSchema.id,
        optionGroupId: productOptionsSchema.optionGroupId,
        name: productOptionsSchema.name,
        additionalPrice: productOptionsSchema.additionalPrice,
        sku: productOptionsSchema.sku,
        stock: productOptionsSchema.stock,
        displayOrder: productOptionsSchema.displayOrder,
      });

    return {
      ...option,
      additionalPrice: Number(option.additionalPrice),
    };
  }

  async updateProductOption(
    optionId: number,
    dto: UpdateProductOptionRequestDto,
  ): Promise<UpdateProductOptionResponseData> {
    const [option] = await this.drizzleService.db
      .update(productOptionsSchema)
      .set({
        name: dto.name,
        additionalPrice: String(dto.additionalPrice),
        sku: dto.sku,
        stock: dto.stock,
        displayOrder: dto.displayOrder,
      })
      .where(and(eq(productOptionsSchema.id, optionId)))
      .returning({
        id: productOptionsSchema.id,
        optionGroupId: productOptionsSchema.optionGroupId,
        name: productOptionsSchema.name,
        additionalPrice: productOptionsSchema.additionalPrice,
        sku: productOptionsSchema.sku,
        stock: productOptionsSchema.stock,
        displayOrder: productOptionsSchema.displayOrder,
      });

    return {
      ...option,
      additionalPrice: Number(option.additionalPrice),
    };
  }

  async deleteProductOption(optionId: number): Promise<void> {
    await this.drizzleService.db
      .delete(productOptionsSchema)
      .where(eq(productOptionsSchema.id, optionId));
  }

  async createProductImage(
    productId: number,
    dto: CreateProductImageRequestDto,
  ): Promise<CreateProductImageResponseData> {
    const [image] = await this.drizzleService.db
      .insert(productImagesSchema)
      .values({
        id: undefined,
        productId,
        url: dto.url,
        altText: dto.altText,
        isPrimary: dto.isPrimary,
        displayOrder: dto.displayOrder,
        optionId: dto.optionId,
      })
      .returning({
        id: productImagesSchema.id,
        url: productImagesSchema.url,
        altText: productImagesSchema.altText,
        isPrimary: productImagesSchema.isPrimary,
        displayOrder: productImagesSchema.displayOrder,
        optionId: productImagesSchema.optionId,
      });

    return image;
  }
}
